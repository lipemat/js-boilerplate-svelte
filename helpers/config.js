"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTypeScriptConfig = getTypeScriptConfig;
const requireJSON5 = require('require-json5');
/**
 * Get the TypeScript config used by Svelte preprocess.
 *
 * Using the config from @tsconfig/svelte/tsconfig.json verbatim.
 *
 * @notice We must use require-json5 because the config uses comments.
 */
function getTypeScriptConfig() {
    return requireJSON5(require.resolve('@tsconfig/svelte/tsconfig.json'));
}
