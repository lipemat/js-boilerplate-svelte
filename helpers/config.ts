const requireJSON5 = require( 'require-json5' );

/**
 * Get the TypeScript config used by Svelte preprocess.
 *
 * Using the config from @tsconfig/svelte/tsconfig.json verbatim.
 *
 * @notice We must use require-json5 because the config uses comments.
 */
export function getTypeScriptConfig() {
	return requireJSON5( require.resolve( '@tsconfig/svelte/tsconfig.json' ) );
}
