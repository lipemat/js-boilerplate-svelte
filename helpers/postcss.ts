import {getConfig} from '@lipemat/js-boilerplate/helpers/config';
import type {CSSModulesOptions, CSSOptions} from 'vite';
import * as postcssScss from 'postcss-scss';
import {getLocalIdent, usingShortCssClasses} from '@lipemat/js-boilerplate/helpers/css-classnames';
import {GetLocalIdent} from 'svelte-preprocess-cssmodules/dist/lib';


type PostCSSConfig = Exclude<CSSOptions['postcss'], string | undefined> & {
	map: boolean;
}

const postcssOptions = getConfig( 'postcss.config' );

const POST_CSS_OPTIONS: PostCSSConfig = {
	plugins: postcssOptions.plugins ?? [],
	map: postcssOptions.sourceMap ?? false,
	parser: postcssScss,
	stringifier: undefined,
	syntax: undefined,
};


export function getGeneratedScopedName(): CSSModulesOptions['generateScopedName'] {
	if ( usingShortCssClasses() && 'production' === process.env.NODE_ENV ) {
		return getShortCssClass;
	}
	return getLocalIdentName();
}


export function maybeGetLocalIdent(): { getLocalIdent: GetLocalIdent } | Record<string, never> {
	if ( ! usingShortCssClasses() || 'production' !== process.env.NODE_ENV ) {
		return {};
	}

	return {
		getLocalIdent: ( context, localIdentName, localName ) => '§' + getLocalIdent( context, localIdentName.interpolatedName, localName ),
	};
}


export function getLocalIdentName(): string {
	return 'production' === process.env.NODE_ENV ? '[contenthash:base52:5]' : '§Ⓜ[name]__[local]__[contenthash:base52:2]';
}

export function getPostCssConfig(): PostCSSConfig {
	return POST_CSS_OPTIONS;
}


function getShortCssClass( filename: string, name: string ): string {
	return '§' + getLocalIdent( {resourcePath: filename ?? ''}, '', name );
}
