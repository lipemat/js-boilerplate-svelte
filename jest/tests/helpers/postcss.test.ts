import {getGeneratedScopedName, getLocalIdentName, getPostCssConfig, maybeGetLocalIdent} from '../../../helpers/postcss';
import {usingShortCssClasses} from '@lipemat/js-boilerplate/helpers/css-classnames';
import type {GetLocalIdent} from 'svelte-preprocess-cssmodules/dist/lib';

// Change this variable during tests.
// noinspection ES6ConvertVarToLetConst
var mockShortCssEnabled = false;

jest.mock( '@lipemat/js-boilerplate/helpers/package-config.js', () => ( {
	...jest.requireActual( '@lipemat/js-boilerplate/helpers/package-config.js' ),
	getPackageConfig: () => ( {
		...jest.requireActual( '@lipemat/js-boilerplate/helpers/package-config.js' ),
		// Change this variable during the test.
		shortCssClasses: mockShortCssEnabled,
	} ),
} ) );

beforeEach( () => {
	mockShortCssEnabled = false;
	process.env.NODE_ENV = 'development';
} );


describe( 'getPostCssConfig', () => {
	it( 'should return a valid PostCSS configuration object', () => {
		const result = getPostCssConfig();
		expect( result ).toMatchSnapshot();
	} );


	it( 'should return a configuration that includes plugins.', () => {
		const result = getPostCssConfig();
		expect( Array.isArray( result.plugins ) ).toBe( true );
		expect( result.plugins ).toHaveLength( 8 );
	} );
} );

describe( 'getLocalIdentName', () => {
	it( 'should return a different string based on the NODE_ENV', () => {
		process.env.NODE_ENV = 'production';
		const prodName = getLocalIdentName();
		expect( prodName ).toBe( '[contenthash:base52:5]' );

		process.env.NODE_ENV = 'development';
		const devName = getLocalIdentName();
		expect( devName ).toBe( '§Ⓜ[name]__[local]__[contenthash:base52:2]' );
	} );


	test( 'Are Short CSS Classes Enabled?', () => {
		expect( usingShortCssClasses() ).toEqual( false );
		mockShortCssEnabled = true;
		expect( usingShortCssClasses() ).toEqual( true );
	} );
} );


describe( 'getGeneratedScopedName', () => {
	it( 'should return a short CSS classname when enabled', () => {
		process.env.NODE_ENV = 'production';
		mockShortCssEnabled = true;
		const shortName = getGeneratedScopedName();
		expect( typeof shortName ).toBe( 'function' );
		if ( 'function' === typeof shortName ) {
			expect( shortName( 'somefile.scss', 'myClass', '' ) ).toMatch( /^§/ );
			expect( shortName( 'somefile.scss', 'myClass', '' ) ).toBe( '§A' );
		}
	} );
} );


describe( 'mabyeGetLocalIdent', () => {
	type LocalIdentParams = Parameters<GetLocalIdent>;

	function makeResource( path: string ): LocalIdentParams[0] {
		return {resourcePath: path} as LocalIdentParams[0];
	}

	function makeIdent( name: string ): LocalIdentParams[1] {
		return {interpolatedName: name} as LocalIdentParams[1];
	}


	it( 'should return an object with getLocalIdent when short CSS classes are enabled', () => {
		mockShortCssEnabled = true;
		process.env.NODE_ENV = 'production';
		const result = maybeGetLocalIdent();
		expect( typeof result ).toBe( 'object' );
		expect( typeof result.getLocalIdent ).toBe( 'function' );
	} );


	it( 'should return an empty object when short CSS classes are not enabled', () => {
		process.env.NODE_ENV = 'development';
		mockShortCssEnabled = true;
		const result = maybeGetLocalIdent();
		expect( typeof result ).toBe( 'object' );
		expect( Object.keys( result ) ).toHaveLength( 0 );

		process.env.NODE_ENV = 'production';
		mockShortCssEnabled = false;
		const result2 = maybeGetLocalIdent();
		expect( typeof result2 ).toBe( 'object' );
		expect( Object.keys( result2 ) ).toHaveLength( 0 );
	} );


	it( 'should return a short CSS classname when enabled', () => {
		process.env.NODE_ENV = 'production';
		mockShortCssEnabled = true;
		const result = maybeGetLocalIdent();
		expect( typeof result ).toBe( 'object' );
		expect( typeof result.getLocalIdent ).toBe( 'function' );


		if ( 'function' === typeof result.getLocalIdent ) {
			expect( result.getLocalIdent( makeResource( 'somefile.scss' ), makeIdent( '' ), 'myClass', {} as LocalIdentParams[3] ) ).toMatch( /^§/ );
			expect( result.getLocalIdent( makeResource( 'somefile.scss' ), makeIdent( '' ), 'myClass', {} as LocalIdentParams[3] ) ).toBe( '§A' );
		}
	} );
} );
