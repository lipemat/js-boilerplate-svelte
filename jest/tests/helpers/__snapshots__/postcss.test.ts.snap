// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`getPostCssConfig should return a valid PostCSS configuration object 1`] = `
{
  "map": true,
  "parser": {
    "default": {
      "parse": [Function],
      "stringify": [Function],
    },
    "parse": [Function],
    "stringify": [Function],
  },
  "plugins": [
    {
      "postcssPlugin": "postcss-global-data",
      "prepare": [Function],
    },
    {
      "Once": [Function],
      "postcssPlugin": "postcss-import",
    },
    [Function],
    [Function],
    {
      "plugins": [
        {
          "postcssPlugin": "postcss-normalize-display-values",
          "prepare": [Function],
        },
        {
          "AtRule": [Function],
          "postcssPlugin": "postcss-font-format-keywords",
        },
        {
          "Declaration": [Function],
          "postcssPlugin": "postcss-oklab-function",
        },
        {
          "postcssPlugin": "postcss-text-decoration-shorthand",
          "prepare": [Function],
        },
        {
          "AtRule": [Function],
          "Once": [Function],
          "postcssPlugin": "postcss-custom-media",
        },
        {
          "Rule": [Function],
          "postcssPlugin": "postcss-nesting",
        },
        {
          "Rule": [Function],
          "postcssPlugin": "postcss-dir-pseudo-class",
        },
        {
          "Declaration": [Function],
          "postcssPlugin": "postcss-lab-function",
        },
        {
          "Rule": [Function],
          "postcssPlugin": "css-blank-pseudo",
        },
        {
          "Declaration": [Function],
          "postcssPlugin": "postcss-page-break",
        },
        {
          "Once": [Function],
          "postcssPlugin": "postcss-font-variant",
        },
        {
          "Declaration": [Function],
          "postcssPlugin": "postcss-stepped-value-functions",
        },
        {
          "browsers": [
            "> 1%",
            "last 1 Android versions",
            "last 1 ChromeAndroid versions",
            "last 2 Chrome versions",
            "last 2 Firefox versions",
            "last 2 Safari versions",
            "last 2 iOS versions",
            "last 2 Edge versions",
            "last 2 Opera versions",
          ],
          "info": [Function],
          "options": {
            "overrideBrowserslist": [
              "> 1%",
              "last 1 Android versions",
              "last 1 ChromeAndroid versions",
              "last 2 Chrome versions",
              "last 2 Firefox versions",
              "last 2 Safari versions",
              "last 2 iOS versions",
              "last 2 Edge versions",
              "last 2 Opera versions",
            ],
          },
          "postcssPlugin": "autoprefixer",
          "prepare": [Function],
        },
        {
          "RuleExit": [Function],
          "postcssPlugin": "postcss-progressive-custom-properties",
        },
        {
          "OnceExit": [Function],
          "postcssPlugin": "postcss-preset-env",
        },
      ],
      "postcssPlugin": "postcss-preset-env",
    },
    [Function],
    {
      "OnceExit": [Function],
      "postcssPlugin": "postcss-sort-media-queries",
    },
    {
      "OnceExit": [Function],
      "postcssPlugin": "js-boilerplate/postcss-pretty",
    },
  ],
  "stringifier": undefined,
  "syntax": undefined,
}
`;
